File: 빙빙 돌며 숫자 사각형 채우기/sqareNumber.py
# 변수 선언 및 입력
n, m = tuple(map(int, input().split()))

# dx, dy 테크닉
dxs, dys = [0,1,0,-1], [1,0,-1,0]
x,y = 0, 0  # 초기위치
dir_num = 0 # R방형
# 격자판 생성
grid = [ 
    [0] * m
    for _ in range(n)
]
# 범위에 포함 되는지 판별
def is_range(nx,ny):
    return 0 <= nx and nx < n and 0 <= ny and ny < m

# 격자판 채우기
for i in range(1, n *m +1):
    grid[x][y] = i

    nx, ny = x + dxs[dir_num] , y + dys[dir_num]

    if not is_range(nx,ny) or grid[nx][ny] != 0:
       dir_num = (dir_num + 1) % 4

    x, y = x + dxs[dir_num] , y + dys[dir_num]

# 출력
for i in range(n):
    for j in range(m):
        print(grid[i][j], end = ' ')
    print()

File: 빙빙 돌며 숫자 사각형 채우기/sqareNumber.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static int n;
    static int m;

    static boolean inRange(int x, int y){
        return 0 <= x && x < n && 0 <= y && y < m;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        int[] dxs = new int[] {0,1,0,-1};
        int[] dys = new int[] {1,0,-1,0};

        int[][] grid = new int[n][m];
        int x  = 0;
        int y = 0;
        int dir_num = 0;

        for(int i = 1; i <= n*m; i++){
            grid[x][y] = i;
            int nx = x + dxs[dir_num];
            int ny = y + dys[dir_num];
            
            if(!inRange(nx,ny) || grid[nx][ny] != 0){
                dir_num = (dir_num + 1) % 4;
            }
            x = x + dxs[dir_num];
            y = y + dys[dir_num];
        }

        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                System.out.print(grid[i][j] + " ");
            }
            System.out.println();
        }
    }
}